tests/unit/test_cli_flags.py

```diff
@@ -349,6 +349,38 @@ def test_log_file_settings_from_config(self):
         assert flags.USE_COLORS is True
         assert flags.USE_COLORS_FILE is False
 
+    @pytest.mark.parametrize(
+        "cli_profile, flag_profile",
+        [
+            (None, None),
+            ("profile", "profile"),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "cli_target, flag_target",
+        [
+            (None, None),
+            ("target", "target"),
+        ],
+    )
+    def test_profile_settings_interaction(
+        self, cli_profile, cli_target, flag_profile, flag_target
+    ):
+        cli_params = ["run"]
+
+        if cli_profile:
+            cli_params = ["--profile", cli_profile] + cli_params
+
+        if cli_target:
+            cli_params = ["--target", cli_target] + cli_params
+
+        context = self.make_dbt_context("run", cli_params)
```

Comment:
To be backwards-compatible, we should make sure that the opposite order works as well.

e.g., we need this to still work:

```shell
dbt run --profile my_profile_name
```

So we'd probably want to make sure this works fine too:

```python
        if cli_profile:
            cli_params = cli_params + ["--profile", cli_profile]

        if cli_target:
            cli_params = cli_params + ["--target", cli_target]
```

Reply:
Sorry for the glacial reply. I've added this scenario to the unit test and it doesn't appear to be working correctly.

I'm failing to get the TARGET and PROFILE flags to be set when they're passed after `run` as you suggest.

Target and profile are added to the global flags and the global flags are in the decorator for the run context in main. The test assertions are failing because `flags.PROFILE` and `flags.TARGET` are always coming back as `None`.

```
@pytest.mark.parametrize(
        "cli_profile, command_profile, flag_profile",
        [
            (None, None, None),
            ("cli_profile", None, "cli_profile"),
            (None, "command_profile", "command_profile"),
            ("cli_profile", "command_profile", "cli_profile"),
        ],
    )
    @pytest.mark.parametrize(
        "cli_target, command_target, flag_target",
        [
            (None, None, None),
            ("cli_target", None, "cli_target"),
            (None, "command_target", "command_target"),
            ("cli_target", "command_target", "cli_target"),
        ],
    )
    def test_profile_settings_interaction_1(
        self, cli_profile, cli_target, command_profile, command_target, flag_profile, flag_target
    ):
        cli_params = ["run"]

        if cli_profile:
            # dbt level --profile flag goes before the run command
            cli_params = ["--profile", cli_profile] + cli_params

        if cli_target:
            # dbt level --target flag goes before the run command
            cli_params = ["--target", cli_target] + cli_params

        if command_profile:
            # command level --profile flag goes after the run command
            cli_params = cli_params + ["--profile", command_profile]

        if command_target:
            # command level --target flag goes after the run command
            cli_params = cli_params + ["--target", command_target]

        context = self.make_dbt_context("run", cli_params)

        flags = Flags(context, None)

        assert flags.PROFILE == flag_profile
        assert flags.TARGET == flag_target
```

Reply:
No worries @barton996

### TLDR
After looking at the pre-existing test cases more closely, we might not even need this at all.

### More detail
i.e., it looks like the general case is already covered by [`test_global_flag_at_child_context`](https://github.com/dbt-labs/dbt-core/blob/c5eb6d70b44f966d17e76e834703bbc6ad05549e/tests/unit/test_cli_flags.py#L356). And the case of specifying it both at the before and after the subcommand is covered by [`test_duplicate_flags_raises_error`](https://github.com/dbt-labs/dbt-core/blob/c5eb6d70b44f966d17e76e834703bbc6ad05549e/tests/unit/test_cli_flags.py#L349). 

So I don't think we need new tests here after all.

If we _do_ end up wanting to add tests specific to `--profile` and `--target`, I think they can be covered in separate test cases since they aren't designed or expected to have interaction effects (unlike the [`test_log_format_interaction`](https://github.com/dbt-labs/dbt-core/blob/c5eb6d70b44f966d17e76e834703bbc6ad05549e/tests/unit/test_cli_flags.py#L296) test that this originally appeared like it was patterned after).

### Potential tests instead

If we do want to add new tests after moving `profile` and `target` to be global flags, here's a couple tests that worked when I tried them out:

```python
    @pytest.mark.parametrize(
        "flag_value",
        [
            (None),
            ("my_profile"),
        ],
    )
    def test_profile_settings(self, flag_value):
        # CLI parameters _before_ subcommand
        parent_context_a = self.make_dbt_context("parent_context_a", ["--profile", flag_value])
        child_context_a = self.make_dbt_context("child_context_a", ["run"], parent_context_a)
        flags_a = Flags(child_context_a)

        # CLI parameters _after_ subcommand
        parent_context_b = self.make_dbt_context("parent_context_b", ["run"])
        child_context_b = self.make_dbt_context(
            "child_context_b", ["--profile", flag_value], parent_context_b
        )
        flags_b = Flags(child_context_b)

        assert flags_a.PROFILE == flags_b.PROFILE
        assert flags_a.PROFILE == flag_value
        assert flags_b.PROFILE == flag_value

    @pytest.mark.parametrize(
        "flag_value",
        [
            (None),
            ("my_target"),
        ],
    )
    def test_target_settings(self, flag_value):
        # CLI parameters _before_ subcommand
        parent_context_a = self.make_dbt_context("parent_context_a", ["--target", flag_value])
        child_context_a = self.make_dbt_context("child_context_a", ["run"], parent_context_a)
        flags_a = Flags(child_context_a)

        # CLI parameters _after_ subcommand
        parent_context_b = self.make_dbt_context("parent_context_b", ["run"])
        child_context_b = self.make_dbt_context(
            "child_context_b", ["--target", flag_value], parent_context_b
        )
        flags_b = Flags(child_context_b)

        assert flags_a.TARGET == flags_b.TARGET
        assert flags_a.TARGET == flag_value
        assert flags_b.TARGET == flag_value
```

Reply:
Per discussion with @peterallenwebb, we feel good about removing this test altogether since:
1. It's covered by [test_global_flag_at_child_context](https://github.com/dbt-labs/dbt-core/blob/c5eb6d70b44f966d17e76e834703bbc6ad05549e/tests/unit/test_cli_flags.py#L356).
2. Doug verified all of this manually also